### PR TITLE
Add MIP-NLP OA/ECP parity baselines

### LINKED ISSUE #112: MIP-NLP OA/ECP parity baseline against MindtPy tests
Parent tracker: https://github.com/bernalde/discopt/issues/110

Branch and PR policy:
- Repository: `bernalde/discopt`.
- Integration branch: `feature/mip-nlp-solver`.
- PR target for this issue: child branch into `feature/mip-nlp-solver` in the fork.
- Final upstream path: one later PR from `bernalde:feature/mip-nlp-solver` into `jkitchin/discopt` after this issue series lands.
- Pyomo MindtPy is a behavior and test reference. Do not copy Pyomo source into discopt without a separate license and attribution review.

References:
- Pyomo MindtPy docs: https://pyomo.readthedocs.io/en/stable/explanation/solvers/mindtpy.html
- Local Pyomo source: ../pyomo/pyomo/contrib/mindtpy
- Convexification-based OA paper: https://arxiv.org/abs/2407.20973

Summary:
Establish native discopt OA/ECP behavior against the small Pyomo MindtPy baseline cases before adding more methods.

Pyomo MindtPy reference tests:
- `../pyomo/pyomo/contrib/mindtpy/tests/test_mindtpy.py`
- `../pyomo/pyomo/contrib/mindtpy/tests/test_mindtpy_ECP.py`
- `../pyomo/pyomo/contrib/mindtpy/tests/MINLP_simple.py`
- `../pyomo/pyomo/contrib/mindtpy/tests/MINLP2_simple.py`
- `../pyomo/pyomo/contrib/mindtpy/tests/test_mindtpy_no_discrete.py`

Implementation notes:
- Re-express the small Pyomo MINLP examples as native discopt models where useful.
- Ensure no-discrete models short-circuit through continuous NLP solve with the MIP-NLP entry point.
- Confirm OA and ECP report objective, bound, gap, status, and incumbent consistently with existing `SolveResult` conventions.
- Fix any current failures in `python/tests/test_oa.py` caused by the new public API.

Acceptance criteria:
- Native discopt tests cover convex OA, ECP, linear objective with nonlinear constraints, no-discrete short-circuit, and infeasible small models.
- Existing OA tests pass through `solver="mip-nlp"` rather than legacy `gdp_method="oa"`.
- Any behavior that intentionally differs from Pyomo is documented in the issue or test comments.

### PR DESCRIPTION
**Pull request by bernalde**
Summary
- Add native discopt versions of the small Pyomo MindtPy APSE and Duran-Grossmann OA/ECP baseline models.
- Exercise both `mip_nlp_method="oa"` and `"ecp"` through `solver="mip-nlp"` and assert objective, bound, gap, status, and incumbent values.
- Tighten no-discrete short-circuit coverage to solve a constrained continuous NLP through the MIP-NLP entry point and assert populated result fields.
- Assert infeasible small-model results do not report an incumbent or gap.

Tests
- `PYTHONPATH=python pytest python/tests/test_oa.py -q`
  - Passed: 7 passed, 14 deselected.
- `PYTHONPATH=python pytest python/tests/test_oa.py -q -m ''`
  - Passed: 21 passed.
- `PYTHONPATH=python pytest python/tests/test_mip_nlp.py -q`
  - Passed: 16 passed.
- `python -m ruff check python/tests/test_oa.py`
  - Passed.
- `python -m ruff format --check python/tests/test_oa.py`
  - Passed.
- `git diff --check`
  - Passed.

Notes
- Local commit hook execution was unavailable because the generated `.git/hooks/pre-commit` points at a missing uv archive interpreter and no `pre-commit` executable or Python module is installed in this shell. The targeted ruff and pytest checks above were run directly.
- Pyomo MindtPy was used only as a behavior reference for the small baseline model equations; no Pyomo source implementation was copied.

Branch Hygiene
- Base branch: `bernalde/discopt:feature/mip-nlp-solver`.
- Head branch: `bernalde/discopt:fix/issue-112-oa-ecp-baseline`.
- Source branch point: `d00ce6c072a59b7455cd535e34a8dd0d505371ab` (`origin/feature/mip-nlp-solver` at branch creation).
- Stacked status: not stacked on another child PR branch; this branch is one commit ahead of `feature/mip-nlp-solver`.
- Prerequisite PRs: depends on the issue-series integration branch `feature/mip-nlp-solver`, which already contains the issue #111 MIP-NLP API routing work.

Closes #112

### CODE REVIEW

**Review by bernalde — COMMENTED** [review]
Blocking: none.

Nonblocking: none.

Questions: none.

Tests run and outcomes:
- Verified PR setup: fetched `origin/feature/mip-nlp-solver` and `origin/fix/issue-112-oa-ecp-baseline`; local merge-base diff matches GitHub stats exactly: 1 file, 85 additions, 5 deletions.
- `PYTHONPATH=python pytest python/tests/test_oa.py -q -m ''`: passed, 21 tests.
- `PYTHONPATH=python pytest python/tests/test_mip_nlp.py -q`: passed, 16 tests.
- `python -m ruff check python/tests/test_oa.py`: passed.
- `python -m ruff format --check python/tests/test_oa.py`: passed.
- `PYTHONPATH=python PYTEST_XDIST_WORKERS=2 make test PYTHON=python PYTEST=pytest`: did not complete locally; xdist workers aborted in unrelated AMP/JAX/Pounce tests under this Python 3.13 shell before reaching a meaningful PR-specific failure. The CI Python fast job for this PR passed on Python 3.12.
- Current CI: Lint + typecheck, Python AMP coverage, Python fast, and Rust passed; Python coverage was skipped by workflow.

Merge-readiness:
- The PR matches issue #112’s requested coverage: native MindtPy-style OA/ECP baselines, no-discrete MIP-NLP short-circuit result fields, infeasible-result conventions, and continued `solver="mip-nlp"` routing.
- I found no actionable issues introduced by this PR. Because this review is submitted from the PR author account, GitHub only permits COMMENT; formal approval must come from another maintainer.

File: python/tests/test_oa.py [hunk 1]
```diff
@@ -122,6 +169,34 @@ def test_linear_objective_nonlinear_constraints(self):
         assert result.x["y"] == pytest.approx(1.0, abs=INTEGRALITY_TOL)
 
 
+@pytest.mark.slow
```
Comment (bernalde) [inline python/tests/test_oa.py]: Nonblocking: this class runs in 1.82s total (slowest case 0.84s), so the slow marker is unwarranted and excludes it from every CI lane. ci.yml runs pytest with -m "not slow ..." in all jobs, and the only slow-inclusive job (amp-integration.yml) targets test_amp_integration.py. As a result these OA/ECP parity baselines — the core deliverable of #112 — never run in CI and give no regression protection. Drop the slow marker so they gate PRs, or add a CI lane that runs the slow test_oa.py tests.

File: python/tests/test_oa.py [hunk 2]
```diff
@@ -182,13 +257,15 @@ class TestOAEdgeCases:
     """Edge cases and degenerate problems."""
 
     def test_pure_nlp_no_integers(self):
```
Comment (bernalde) [inline python/tests/test_oa.py]: Nonblocking: the function is still named test_pure_nlp_no_integers but now builds model "no_discrete_short_circuit" with a constraint (x**2 >= 1, minimize x). The behavior change is fine, but rename the test (e.g. test_no_discrete_short_circuit) so the name matches the model and docstring for discoverability.

**Review by bernalde — COMMENTED** [review]
Senior-maintainer review of #123 (Add MIP-NLP OA/ECP parity baselines).

Reviewed at head 37c8ec98 against merge-base d00ce6c (base feature/mip-nlp-solver). Diff confirmed: 1 file, +85/-5 — matches the PR's reported totals. Test-only change to python/tests/test_oa.py.

Verdict note: I am the PR author, so GitHub only permits a COMMENT review. There are no blocking issues; the formal APPROVE must come from another maintainer.

## Intent vs. issue #112

The PR delivers the issue's scope: native discopt versions of the small MindtPy APSE and Duran-Grossmann models, exercised through solver="mip-nlp" for both mip_nlp_method="oa" and "ecp", plus the no-discrete short-circuit and infeasible cases. I verified the asserted optima are correct, not coincidental: simple MINLP obj=3.5 at y=[0,1,0] (x=(1,1): 1.5 + 1 + 1 = 3.5), Duran-Grossmann obj=6.00976 at y=[0,1,0] — both match the known MindtPy optima, and bound==objective with gap=0.0 for all four solves. Closes #112 is appropriate since the base is the integration branch and the acceptance criteria are covered.

## Nonblocking

1. The parity baselines are marked @pytest.mark.slow but are not slow, so they run in no CI lane. Measured: the whole TestMindtPyBaselineParity class runs in 1.82s (slowest single case 0.84s). Every CI job in .github/workflows/ci.yml runs pytest with -m "not slow and not correctness and not integration ...", and the only slow-inclusive job (amp-integration.yml) targets test_amp_integration.py, not test_oa.py. Net effect: the headline deliverable of #112 — the OA/ECP convex-parity and linear-objective/nonlinear-constraint coverage — provides zero automated regression protection and only runs on a manual local slow run. Fix: drop the slow marker from TestMindtPyBaselineParity (these sub-second convex MINLPs belong in the PR-fast lane), or add a CI job that runs the slow test_oa.py lane. The modified no-discrete and infeasible edge-case tests are not slow-marked and do gate CI, so this applies only to the new class.

2. test_pure_nlp_no_integers no longer matches its name. It now builds model "no_discrete_short_circuit" with a constraint (x**2 >= 1, minimize x -> x=1), not a pure unconstrained NLP. The behavior change is fine and arguably better, but the stale function name hurts discoverability. Rename the test (e.g. test_no_discrete_short_circuit) to match the model and docstring.

## Questions

None.

## Minor (non-actionable note)

The assertions gap == approx(0.0, abs=1e-9) and the exact incumbent np.asarray(result.x["y"]).tolist() == approx([0.0, 1.0, 0.0]) are tight. They are correct and robust for these unique-optimum convex baselines (verified gap is exactly 0.0), but if this class is later extended to problems with alternate optima or a nonzero converged gap, prefer asserting gap <= gap_tolerance and relaxing the exact-incumbent check. No change needed now.

## Tests run (clean checkout at 37c8ec98, .venv)

- pytest python/tests/test_oa.py -o addopts="" -q -> 21 passed (65s; the bulk is pre-existing slow tests, not the new class).
- pytest python/tests/test_oa.py::TestMindtPyBaselineParity -o addopts="" --durations=10 -> 4 passed in 1.82s.
- Re-solved both models for oa and ecp directly: obj 3.5 / 6.00976, bound==obj, gap=0.0, y=[0,1,0] — assertions are accurate.
- ruff check and ruff format --check on test_oa.py -> passed.

## Merge readiness

No blocking issues; the tests are correct, the expected values are right, and the issue's acceptance criteria are covered. The one substantive item is Nonblocking: the parity class is excluded from CI by an unwarranted slow marker, which undercuts its purpose as a regression baseline — worth fixing before merge but not a correctness blocker. The formal verdict must come from another maintainer since I authored this PR.

**Comment by bernalde** [comment]
Pushed commit:
- d73323b8d0bbeee67ec3150cc052fdd7373fac46: test: address OA review comments
  https://github.com/bernalde/discopt/commit/d73323b8d0bbeee67ec3150cc052fdd7373fac46

Main changes:
- Removed the `slow` marker from `TestMindtPyBaselineParity` so the OA/ECP parity baseline tests run in the default `test_oa.py` path and can gate PRs.
- Renamed `test_pure_nlp_no_integers` to `test_no_discrete_short_circuit` to match the constrained no-discrete MIP-NLP model and docstring.

Verification:
- `PYTHONPATH=python pytest python/tests/test_oa.py -q`: passed, 11 passed and 10 deselected.
- `PYTHONPATH=python pytest python/tests/test_mip_nlp.py -q`: passed, 16 passed.
- `PYTHONPATH=python pytest python/tests/test_oa.py -q -m ''`: passed, 21 passed.
- `python -m ruff check python/tests/test_oa.py`: passed.
- `python -m ruff format --check python/tests/test_oa.py`: passed.
- `PYTHONPATH=python make lint PYTHON=python`: passed, `ruff check python/` and `ruff format --check python/`.

Type-check status:
- `pre-commit` is not installed in this local environment, so the local commit hook could not run.
- Direct mypy with the local Python 3.13 environment failed before checking project code because the installed NumPy 2.5 stubs require newer syntax than the repo's configured Python 3.10 target.
- `python3.12 -m mypy ...` could not run because mypy is not installed for that interpreter.
- `uv run python -m mypy python/discopt --ignore-missing-imports --python-version 3.12` reached project code and failed on existing unrelated type errors in `python/discopt/nn/tree.py`, `python/discopt/export/nl.py`, and `python/discopt/estimate.py`.

Comments intentionally not addressed:
- None.

Remaining risks or follow-up:
- GitHub Actions checks for the pushed commit were still running when I checked.

File: python/tests/test_oa.py (hunk 1 above)
Comment (bernalde) [reply python/tests/test_oa.py]: Addressed in https://github.com/bernalde/discopt/commit/d73323b8d0bbeee67ec3150cc052fdd7373fac46: removed the slow marker from TestMindtPyBaselineParity so these OA/ECP parity baselines run in the default test_oa.py path. Verified with PYTHONPATH=python pytest python/tests/test_oa.py -q.

File: python/tests/test_oa.py (hunk 2 above)
Comment (bernalde) [reply python/tests/test_oa.py]: Addressed in https://github.com/bernalde/discopt/commit/d73323b8d0bbeee67ec3150cc052fdd7373fac46: renamed the test to test_no_discrete_short_circuit so the name matches the constrained no-discrete MIP-NLP case. Verified with PYTHONPATH=python pytest python/tests/test_oa.py -q.
